arroyo/processing/strategies/dead_letter_queue/policies/produce.py

```diff
@@ -0,0 +1,71 @@
+import json
+import time
+from collections import deque
+from concurrent.futures import Future
+from typing import Deque, Optional
+
+from arroyo.backends.abstract import Producer
+from arroyo.backends.kafka.consumer import KafkaPayload
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessage,
+    InvalidMessages,
+)
+from arroyo.types import Message, Topic
+from arroyo.utils.metrics import get_metrics
+
+MAX_QUEUE_SIZE = 5000
+
+
+class ProduceInvalidMessagePolicy(DeadLetterQueuePolicy):
+    """
+    Produces given InvalidMessages to a dead letter topic.
+    """
+
+    def __init__(
+        self, producer: Producer[KafkaPayload], dead_letter_topic: Topic
+    ) -> None:
+        self.__metrics = get_metrics()
+        self.__dead_letter_topic = dead_letter_topic
+        self.__producer = producer
+        self.__futures: Deque[Future[Message[KafkaPayload]]] = deque()
+
+    def handle_invalid_messages(self, e: InvalidMessages) -> None:
+        """
+        Produces a message to the given dead letter topic for each
+        invalid message. Produced message is in the form provided
+        by `InvalidMessage.to_dict()`
+        """
+        for message in e.messages:
+            payload = self._build_payload(message)
+            self._produce(payload)
+        self.__metrics.increment("dlq.produced_messages", len(e.messages))
+
+    def _build_payload(self, message: InvalidMessage) -> KafkaPayload:
+        data = json.dumps(message.to_dict()).encode("utf-8")
+        return KafkaPayload(key=None, value=data, headers=[])
+
+    def _produce(self, payload: KafkaPayload) -> None:
+        """
+        Prune done futures from queue if filled, then asynchronously
+        produce the message, adding the process (future) to the queue.
+        """
+        self._prune_done_future()
+        self.__futures.append(
+            self.__producer.produce(
+                destination=self.__dead_letter_topic, payload=payload
+            )
+        )
+
+    def _prune_done_future(self) -> None:
+        if len(self.__futures) >= MAX_QUEUE_SIZE:
```

Comment:
Please remove all the done elements from the future queue.
Basically remove all the futures at the beginning of the queue where `done()` returns true.